service/lib/agama/storage/proposal_settings_reader.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+# Copyright (c) [2023] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "y2storage/encryption_method"
+require "y2storage/pbkd_function"
+require "agama/storage/volume_templates_builder"
+require "agama/storage/space_settings"
+
+module Agama
+  module Storage
+    # Proposal settings reader.
+    class ProposalSettingsReader
+      # @param config [Agama::Config]
+      def initialize(config)
+        @config = config
+      end
+
+      # Reads the proposal settings from the control file.
+      #
+      # @return [ProposalSettings]
+      def read
+        ProposalSettings.new.tap do |settings|
+          config.data.fetch("storage", {}).each do |key, value|
+            reader = READERS[key]
+            send(reader, settings, value) if reader
```

Comment:
what mean if there is no reader? Does it make sense to log that key does not have reader?

Reply:
No reader means there is an unknown key in the config file. Again, all these cases fall down into the missing validation part that will be added later.